galter_subjects_utils/writer.py

```diff
@@ -26,3 +29,89 @@ def write_jsonl(
             f.write("\n")
 
     return filepath
+
+
+class SubjectDeltaLogger:
+    """Convenience logger for delta operations applied to records."""
+
+    def __init__(self, filepath=None):
+        """Constructor."""
+        if not filepath:
+            # In memory
+            self.f = StringIO()
+        else:
+            self.f = open(filepath, "w+", newline='')
+
+        self.header = ["pid", "time", "error", "deltas"]
+        self.writer = csv.DictWriter(
+            self.f, fieldnames=self.header
+        )
+        self.writer.writeheader()
+
+        self.pid = None
+        self.error = None
+        self.deltas = None
+
+
+    def log(self, pid, delta=None, error=None):
+        """Buffered log.
+
+        Idea is to build up a log entry for pid with multiple deltas.
+        The actual logging is only written out when flush() is called.
+        An impromptu change in pid will clear out any accumulated data.
+        `log(error=...)` is typically called on its own to fill the error
+        portion.
+        """
+        if self.pid != pid:
+            self.clear()
+
+        self.pid = pid
+
+        if delta:
+            if delta["type"] == "replace":
+                delta_msg = f"{delta['id']} -> {delta['new_id']}"
+            elif delta["type"] == "remove":
+                delta_msg = f"{delta['id']} -> X"
```

Comment:
Admittedly "cute" representation of changes for human consumption. Meh :shrug: .